Point-like IRFs

TarekHC opened the pull request:
Dear all, 

This is a first draft of the proposal Christoph and I prepared to include the point-like IRF format. The main motivation is to allow within the DL3 format the approach to generate IRFs used by the current generation of IACTs.

You can find an updated version in:
http://www.gae.ucm.es/~thassan/gamma/irfs/index.html

The IRF section was re-organized: IRF components are now described in a dedicated separated section

We can discuss this PR during the open DL3 call and also during the MPIK meeting.

kosack commented:
is it necessary to have `_full` or `_point` appended to all tables if the `hdu_clas2` already encodes that fact that it is point-like? Isn't that redundant  and makes a "generic" code to load it more difficult?

TarekHC commented:
> is it necessary to have `_full` or `_point` appended to all tables if the `hdu_clas2` already encodes that fact that it is point-like? Isn't that redundant  and makes a "generic" code to load it more difficult?

I agree. Although, for now, we still do not have a very clear HDU_CLASn hierarchy defined. That is the reason why I was not very cautious with the HDU_CLAS naming (e. g. "bkg_2d", should probably also be divided into "Bkg" and "2D"). We could either wait for a dedicated PR, or I could implement a first approach for the naming convention here, although this might trigger discussion not associated with the new proposed format.

My proposal for HDU_CLAS1 -> HDU_CLAS4 would be (the labels could be changed, focus on the hierarchy):
```
RESPONSE  
        FULL-ENCLOSURE
                AEFF
                        2D
                BKG 
                        2D
                        3D
....               
        POINT-LIKE
                AEFF
                        2D
                BKG 
                        2D
                        3D
....
```

cdeil commented:
@TarekHC - Thanks!

This PR is the result of #71 and offline discussions.

We will discuss it today at 11 am in the telcon, as well as in the f2f next week.
See https://github.com/open-gamma-ray-astro/2017-03_IACT_DL3_Meeting

People that are involved in IRF production with CTA / HESS / ... or consumption by science tools (Gammapy, ctools) -- please have a look and send comments!
The goal here is to try and agree on something and merge by the end of next week latest.

cc @jknodlseder @jjlk @gernotmaier @bkhelifi @registerrier 

I already made some comments in  #71,  and I will have a closer look at this PR in the coming days.

My feeling at the moment is that:
- the definition of the point-like responses by themselves is pretty straightforward, it's mostly a copy of the full-enclosure responses, with header keys declaring what they are and minor additions (the THETA cut info e.g.)
- there's no concept here how to link EVENTS to IRFs, proposing solutions for that is future work (e.g. updates to the HDU index table or observation definition files), probably based on prototying experience in science tools.
- The HDUCLAS keys will most likely be part of the solution how to locate and interpret IRFs correctly by science tools. Given that we have no concept how they link to EVENTS yet, it's difficult to decide what to put. It's also difficult to decide what piece of info to put at which level of HDUCLAS. My understanding is that so far we didn't really have useful info about HDUCLAS in the spec, and there's no DL3 producers or tools yet, so backwards compatibility is not a concern here.

So overall I'm +1 to make this change & addition to the DL3 spec within the next ~ week as a step in the right direction. It *is* useful because it gives DL3 producers info what to put, and science tool prototyping to use HDUCLAS and to do analyses with both point-like and full-enclosure responses can start. But the understanding is that this will be what we use in the coming month, with the hope that by the end of the year we have a more complete scheme developed (either an evolution of what we're doing here, or just something entirely new that supersedes what we do here).

TarekHC commented:
As discussed @ Heidelberg, I added a general section introducing the HDUCLASn hierarchy, and added HDUCLASn keywords to all IRFs.

We will probably modify the names of current IRF components format, and even the HDUCLASn labels, but the hierarchy will probably be the one proposed here:

http://www.gae.ucm.es/~thassan/gamma/general/hdu-hierarchy.html

cdeil commented on source/general/hdu-hierarchy.rst:
```diff
@@ -0,0 +1,87 @@
+* HDUCLAS3: In case of ``RESPONSE`` type, refers to the way the IRF component was produced (``POINT-LIKE`` or ``FULL-ENCLOSURE``)
+
+* HDUCLAS4: In case of ``RESPONSE`` type, refers to the name of the specific format
+
```
The description of what to put for HDUDOC could be central, putting the Github URL.
For HDUVERS I would suggest to put "0.2" or "1.0" for now for all formats, and then on the general page describe that they can evolve differently, that it's not coupled to the global version from this spec.

TarekHC replied:
Added a section describing briefly the different types of HDUs in the hduclass section.
Added a reference to that section in the definition of HDUCLASS keywords.
Currently we just define HDUDOC and HDUVERS in this section. I would try to de-correlate this versioning discussion with this PR... Or we will never merge.

TarekHC commented on source/general/hdu-hierarchy.rst:
```diff
@@ -0,0 +1,87 @@
+
+.. _hdu-hierarchy:
+
+HDU hierarchy
```
Fixed

TarekHC commented on source/general/hdu-hierarchy.rst:
```diff
@@ -0,0 +1,87 @@
++------------+-----------+------------+-----------------+--------------+
+|            |  EVENTS   |            |                 |              |
++------------+-----------+------------+-----------------+--------------+
+|            |  RESPONSE | EFF_AREA   |   POINT-LIKE    |   aeff_2d    |
```
Argh, that was already corrected, and I did not push the commits... Now it should be correct.

TarekHC commented:
Hi @cdeil ,

Thank you for the review!

I believe I incorporated all your suggestions. There were some that I had fixed locally during the meeting but forgot to push... (yes, I know, it's annoying... sorry about that!)

I also reformatted the point-like section as you suggested. As you said, I have the feeling that when somebody is taking a look at the specifications of a format, they really just want to see the whole thing together, without needing to go through different sections to understand the little differences. But of course for us it would be a pain to maintain both sections, so let me know if you like the new format.

I also added "True" or "Reconstructed" to all energy axes (following discussion in #63 ). The labelling is still terribly inconsistent (background ``ENERGY`` is reconstructed, while in the effective area is ``ENERGY`` refers to true energy and ``ERECO`` to the reconstructed one... a mess!).

You can find the updated specs online at:
http://www.gae.ucm.es/~thassan/gamma/index.html

cdeil commented:
@TarekHC - Thanks!

I see that now e.g. for EDISP the page
http://www.gae.ucm.es/~thassan/gamma/irfs/full_enclosure/energy_dispersion/index.html
is separate from the description of EDISP at
http://www.gae.ucm.es/~thassan/gamma/irfs/irf_components/index.html#energy-dispersion

@TarekHC - Wouldn't it be better to keep them together?

If you want to keep as-is, please add links from the pages under `irfs/full_enclosure` to the sections in `irfs/irf_components`. (currently there are statements like "Energy migration axis (defined above)", but it's no longer "above" but on a separate page now).

TarekHC commented:
I think that is probably just a matter of taste. 

I would go for leaving as it is (and adding references, as you suggest). I think separating the IRF component "formal" definitions from the description of the format(s) (technical and more "pragmatic") makes sense. Although others may not like it, so I'm totally open to both possibilities.

I will add the references now, and if anyone has a strong opinion, then I will change it back.

TarekHC commented:
I think that is probably just a matter of taste. 

I would go for leaving as it is (and adding references, as you suggest). I think separating the IRF component "formal" definitions from the description of the format(s) (technical and more "pragmatic") makes sense. Although others may not like it, so I'm totally open to both possibilities.

I added the references, and if anyone has a strong opinion, then I will change it back.

cdeil commented:
@TarekHC - Agreed that structuring of IRF description is a matter of taste and both are valid ways to put it. Merging as-is now. Thanks for all your work on this pull request and patience with the review!
